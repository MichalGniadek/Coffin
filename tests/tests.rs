use coffin2::{
    ast::Ast,
    debug_print::DebugPrint,
    error::CoffinError,
    lexer, name_resolution,
    parser::{self, spans_table::SpansTable},
    type_resolution,
};
use insta::{assert_snapshot, glob};
use lasso::RodeoResolver;
use std::{fs, path::Path};

fn get_ast(path: &Path) -> (Ast, SpansTable, RodeoResolver, Vec<CoffinError>) {
    let code = fs::read_to_string(path).unwrap();
    let lexer = lexer::lex(&code);
    parser::parse(lexer)
}

#[test]
fn insta() {
    glob!(r"insta_parser\*.coff", |path| {
        let (ast, _, rodeo, errors) = get_ast(path);
        assert_snapshot!(DebugPrint::visit(
            &ast,
            Some(&rodeo),
            None,
            None,
            None,
            Some(errors)
        ));
    });
    glob!(r"insta_parser_spans\*.coff", |path| {
        let (ast, spans, rodeo, errors) = get_ast(path);
        assert_snapshot!(DebugPrint::visit(
            &ast,
            Some(&rodeo),
            Some(&spans),
            None,
            None,
            Some(errors)
        ));
    });
    glob!(r"insta_variables\*.coff", |path| {
        let (ast, spans, rodeo, mut errors0) = get_ast(path);
        let (vars, errors1) = name_resolution::visit(&ast, &spans);
        errors0.extend(errors1);
        assert_snapshot!(DebugPrint::visit(
            &ast,
            Some(&rodeo),
            None,
            Some(&vars),
            None,
            Some(errors0)
        ));
    });
    glob!(r"insta_types\*.coff", |path| {
        let (ast, spans, rodeo, mut errors0) = get_ast(path);
        let (vars, errors1) = name_resolution::visit(&ast, &spans);
        let types = type_resolution::visit(&ast, &vars, &rodeo);
        errors0.extend(errors1);
        assert_snapshot!(DebugPrint::visit(
            &ast,
            Some(&rodeo),
            None,
            None,
            Some(&types),
            Some(errors0)
        ));
    });
}
