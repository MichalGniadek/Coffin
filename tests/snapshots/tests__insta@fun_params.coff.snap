---
source: tests/tests.rs
expression: "DebugPrint::visit(&ast, Some(&rodeo), None, None, Some(&types), Some(errors0))"

---
fun asd (a : int, b : int)  {
	(a = #3: int): void
	(b = #5: int): void
}: void

===
